### DISCUSSION TITLE
Implement WebWorker for meshopt_decoder.js

### DISCUSSION THREAD

**Discussion by Ben-Mack** [zeux/meshoptimizer#253] [Ideas]
Thanks for the great works on this lib!
It would be very nice if meshopt_decoder.js could implement web worker support (like what BasisLoader did).
It gonna improve decoding performance and reduce freezing a lot, especially on a large file.

**Comment by zeux** [comment]
Out of curiosity, do you have a specific use case in mind or are you just wondering if this would be helpful? The reason why I ask is I've thought about providing some sort of helper as part of meshopt_decoder.js to make WebWorker integration easier, but so far it hasn't been necessary - meshopt_decoder runs circles around Basis transcoder and on all data I've tested it usually didn't result in noticeable pauses during glTF file loading, esp. when using Chrome versions that have Wasm SIMD enabled. If you have examples to the contrary I'd be happy to look at them. It should definitely be possible to implement webworkers on top of the decoder .js today, it's just that it's not integrated and so far I haven't seen the need yet.

**Reply by wallabyway** [reply]
Hi @zeux - yes, I can provide an example of freezing.  It's a large cad model... I'll ping you directly on glTF / Slack with a link, so you can test it yourself.

As you say, the wasm SIMD decoder is already extremely fast, but for some reason, there is still some freezing in the UI on large decodes.  

I haven't profiled the decoder, so I'm not sure if the freezing is from memory transfer after the decoder processing, or decoder processing on the main thread.  However, freezing is noticeable.  

Note: In my demo, I hide the UI freezing, by cascade loading tricks, which causes 'stuttering' but, is less noticeable than a complete UI freeze.
Hopefully, switching to a web-worker fixes the problem without complicating your great implementation or bloating the LOC.

**Comment by zeux** [comment]
Just an update to close this out, but this has been implemented and is available in meshoptimizer 0.18.
